### PR TITLE
feat(notifier): avoids multiple offers creation per provider

### PR DESCRIPTION
**Pull request by itofarina**
## Features
- [x] Add wrapper to sell page and enable to register only if check with cache that there are no offers as provider from the current account
- [x] Show spinner while fetching offers for current provider
- [x] Show "no multiple offers" card for notifier provider with redirection to your offer for edition
## Demo

https://user-images.githubusercontent.com/8449567/121590602-e24d4600-ca0e-11eb-9322-4176a1c9bb6f.mov

### CODE REVIEW

File: src/components/pages/notifier/sell/PageWrapper.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        provider: account,
+      })
+      setHasOffer(Boolean(currentOwnOffers[0]))
+      setIsLoadingOffer(false)
+    }
+
+    getOwnOffers()
+  }, [apis, account, reportError])
+
+  if (hasOffer) {
```
Comment (vjgene) [inline src/components/pages/notifier/sell/PageWrapper.tsx]: Can we make another condition that if provider tries to register the same url that it shows a message you are already registered as a provider ?

File: src/api/rif-marketplace-cache/notifier/offers/__tests__/api.test.ts [hunk 2]
```diff
@@ -77,12 +88,12 @@ describe('Notifier Offers Service', () => {
...
+      await offersService.fetch(filters)
       expect(serviceFindSpy).toBeCalled()
     })
 
     test('should return NotifierOfferItem[] on success', async () => {
-      const actualOffers: NotifierOfferItem[] = await offersService.fetch()
+      const actualOffers: NotifierOfferItem[] = await offersService.fetch(filters)
 
       const expectedOffers: NotifierOfferItem[] = MOCK_RESPONSE
         .map(mapFromTransport)
```
Comment (vjgene) [inline src/api/rif-marketplace-cache/notifier/offers/__tests__/api.test.ts:99]: can we use a better name than mapFromTransport? this seems to be used in multiple places.

File: src/api/rif-marketplace-cache/notifier/offers/api.ts [hunk 3]
```diff
@@ -56,13 +56,13 @@ class OffersService extends AbstractAPIService
...
-        ...filters,
-        currency: Array.from(filters.currency),
-        channels: Array.from(filters.channels),
-      },
-    })
+    const { currency, channels } = filters
+    const query = filters && {
+      ...filters,
+      currency: currency ? Array.from(currency) : undefined,
+      channels: channels ? Array.from(channels) : undefined,
```
Comment (vjgene) [inline src/api/rif-marketplace-cache/notifier/offers/api.ts]: why is undefined being used. if it's an array can we not use []?

**Review by vjgene — CHANGES_REQUESTED** [review]
Looks good overall. I just think it'd be more clear from provider perspective  to add an extra message when provider tries to register using same url again for clarity.

**Comment by itofarina** [comment]
@vjgene thanks for reviewing. This PR is still a work in progress as it's a draft 🙂

File: src/api/rif-marketplace-cache/notifier/offers/api.ts (hunk 3 above)
Comment (itofarina) [reply src/api/rif-marketplace-cache/notifier/offers/api.ts]: I think we should be able to only specify the needed filters. Like it's better to not specify the currency if you don't want to filter by currency. Does it make sense?

File: src/api/rif-marketplace-cache/notifier/offers/__tests__/api.test.ts (hunk 2 above)
Comment (itofarina) [reply src/api/rif-marketplace-cache/notifier/offers/__tests__/api.test.ts:99]: I think this is not related to this PR. Personally, I like this name like it's a standard to parse the DTOs. We can discuss about changing it of course but don't think this PR is the best place to do it 😄.

File: src/components/pages/notifier/sell/PageWrapper.tsx (hunk 1 above)
Comment (itofarina) [reply src/components/pages/notifier/sell/PageWrapper.tsx]: Actually, if the provider already has an offer we don't show the form, so they won't be able to enter a url.

File: src/components/organisms/storage/sell/NoMultipleOffersCard.tsx [hunk 4]
```diff
@@ -15,13 +15,22 @@ const useStyles = makeStyles(() => ({
...
+const MY_OFFERS_ROUTE: Record<Service, string> = {
+  storage: ROUTES.STORAGE.MYOFFERS.BASE,
+  notifier: ROUTES.NOTIFIER.MYOFFERS.BASE,
+}
+
+type Props = {
+  service: Service
+}
+
+const NoMultipleOffersCard: FC<Props> = ({ service }) => {
```
Comment (jurajpiar) [inline src/components/organisms/storage/sell/NoMultipleOffersCard.tsx:29]: Take it out of the storage folder if it is meant to be reused outside of storage.
Also, it seems rather hacky. Maybe would be better to take path as argument pass route to it directly?

File: src/components/pages/notifier/sell/NotifierSellPage.tsx [hunk 5]
```diff
@@ -29,6 +33,25 @@ const NotifierSellPage: FC = () => {
...
+    const getOwnOffers = async (): Promise<void> => {
+      setIsLoadingOffer(true)
+
+      const offersService = apis['notifier/v0/offers'] as NotifierOffersService
+      offersService.connect(reportError)
+
+      const currentOwnOffers = await offersService.fetch({
+        provider: account,
+      })
+      setHasOffer(Boolean(currentOwnOffers[0]))
```
Comment (jurajpiar) [inline src/components/pages/notifier/sell/NotifierSellPage.tsx]: Might be safer for runtime to do:
```suggestion
      setHasOffer(Boolean(currentOwnOffers?.length))
```

File: src/models/marketItems/NotifierFilters.ts [hunk 6]
```diff
@@ -10,8 +10,8 @@ type PriceFilter = MinMaxFilter & {
 export type NotifierOffersFilters = {
     size: MinMaxFilter
     price: PriceFilter
-    currency: Set<SupportedTokenSymbol>
-    channels: Set<string>
+    currency?: Set<SupportedTokenSymbol>
+    channels?: Set<string>
```
Comment (jurajpiar) [inline src/models/marketItems/NotifierFilters.ts:14]: Let me fix the type of the api's _fetch sig, but I think we should keep these enforced as they serve the purpose (probably) and instead type cast upon use.

**Review by jurajpiar — CHANGES_REQUESTED** [review]

**Review by jurajpiar — APPROVED** [review]

File: src/components/organisms/storage/sell/NoMultipleOffersCard.tsx (hunk 4 above)
Comment (itofarina) [reply src/components/organisms/storage/sell/NoMultipleOffersCard.tsx:29]: Moved, thanks

File: src/components/pages/notifier/sell/NotifierSellPage.tsx [hunk 7]
```diff
@@ -29,6 +34,25 @@ const NotifierSellPage: FC = () => {
...
+  useEffect(() => {
+    const getOwnOffers = async (): Promise<void> => {
+      setIsLoadingOffer(true)
+
+      const offersService = apis['notifier/v0/offers'] as NotifierCacheAPIService
+      offersService.connect(reportError)
+
+      const currentOwnOffers = await offersService.fetch({
+        provider: account,
+      } as NotifierOffersFilters)
```
Comment (vjgene) [inline src/components/pages/notifier/sell/NotifierSellPage.tsx:49]: What happens if it is the same url but a different provider account? should we also block registration if the url is already registered?

**Review by vjgene — APPROVED** [review]
Everything looks good. Just left a note on registering with same url.

File: src/components/pages/notifier/sell/NotifierSellPage.tsx (hunk 7 above)
Comment (itofarina) [reply src/components/pages/notifier/sell/NotifierSellPage.tsx:49]: Thanks, that's a good point! I might be missing something but why you wouldn't be able to provide the same service with another account?
At least in the SC, the only check is against the account ([reference](https://github.com/rsksmart/rif-marketplace-notifier/blob/master/contracts/NotifierManager.sol#L153))
May be @artem-brazhnikov could help us define this behavior?

Comment (itofarina) [reply src/components/pages/notifier/sell/NotifierSellPage.tsx:49]: Resolving as it will be tackled on a separated story

**Pull request merged**